File: 371-SumOfTwoIntegers/SumofTwoIntegers.java
public class Solution {
    public int getSum(int a, int b) {
        // a + b = 2* (a&b) + a^b
        // LHS -> RHS
        // If a&b = 0, then finished.
        // Else, number of 1 (in the binary representation) decreases strictly.
        if (a == 0 ) return b;
        // if (b == 0) return a;
        return getSum((a & b) << 1, a ^ b);
    }
}

File: 371-SumOfTwoIntegers/SumofTwoIntegers.cpp
class Solution {
public:
    int getSum(int a, int b) {
        while(1){
            int nocarry=a^b;
            int carry=(a&b)<<1;
            if(carry==0)    return nocarry;
            return getSum(carry,nocarry);
        }
    }
};